pkg/reconciler/managed/reconciler.go

```diff
@@ -1248,3 +1291,92 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (resu
+	record event.Recorder, opType protov1alpha1.OperationType, changeErr error, ad *AdditionalDetails,
+) {
+	if !r.features.Enabled(feature.EnableAlphaChangeLogs) {
+		// change log feature isn't enabled, just return immediately
+		return
+	}
+
+	// convert any additional details to a Struct to include in the change log entry
+	var ads *structpb.Struct
+	if ad != nil {
+		b, err := json.Marshal(ad)
+		if err != nil {
+			log.Debug("Cannot marshal additional details", "error", err)
+			record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+		} else {
+			ads = &structpb.Struct{}
+			if err = ads.UnmarshalJSON(b); err != nil {
+				log.Debug("Cannot convert additional details to Struct", "error", err)
+				record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+				ads = nil
+			}
+		}
+	}
+
+	// determine the full namespaced name of the managed resource
+	var namespacedName string
+	namespace := managed.GetNamespace()
+	if namespace != "" {
+		namespacedName = fmt.Sprintf("%s/%s", namespace, managed.GetName())
+	} else {
+		namespacedName = managed.GetName()
+	}
+
+	// get an error message from the error if it exists
+	var changeErrMessage *string
+	if changeErr != nil {
+		changeErrMessage = reference.ToPtrValue(changeErr.Error())
+	}
+
+	// send everything we've got to the change log service
+	cle := &protov1alpha1.ChangeLogEntry{
+		Provider:          r.providerVersion,
+		Type:              managed.GetObjectKind().GroupVersionKind().String(),
+		Name:              namespacedName,
+		ExternalName:      meta.GetExternalName(managed),
+		Operation:         opType,
+		Snapshot:          snapshot,
+		ErrorMessage:      changeErrMessage,
+		AdditionalDetails: ads,
+	}
+
+	clr, err := r.changeLogClient.SendChangeLog(ctx, cle)
+	if err != nil {
+		log.Debug("Cannot send change log entry", "error", err)
+		record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
+		return
+	}
+
+	log.Debug("Sent change log entry", "changeLogResponse", clr)
+}
+
+// takeChangeLogSnapshot attempts to take a snapshot of the given managed
+// resource by converting it to a Struct that will be sent to the change logs
+// after an external operation is performed, if the change logs feature is
+// enabled. If there is an error during this process, an event will be recorded
+// and a nil snapshot will be returned.
+func (r *Reconciler) takeChangeLogSnapshot(managed resource.Managed, log logging.Logger, record event.Recorder) *structpb.Struct {
+	if !r.features.Enabled(feature.EnableAlphaChangeLogs) {
+		// change log feature isn't enabled, just return immediately
+		return nil
+	}
+
+	// capture the full state of the managed resource from before we performed the change
+	snapshot, err := resource.AsStruct(managed)
+	if err != nil {
+		log.Debug("Cannot snapshot managed resource", "error", err)
+		record.Event(managed, event.Warning(reasonCannotSendChangeLog, err))
```

Comment:
Are events the right vehicle to report such an error?

Reply:
Great question @sttts! I'd love to hear your opinion if you have one 🙇‍♂️ 

I was thinking that the core reconciliation of the managed resource is the most important thing, and if the change log fails to send then I'd still like reconciliation to complete. A (somewhat) visible way to note that an error happened, but to not alter the rest of the reconciliation attempt, was to just surface it as a warning event on the resource.

Do you have an opinion or other ideas?

Reply:
Fearing lots of such events when something goes wrong with the audit log collector.

Reply:
Alternative: log error and emit metrics.

Reply:
@negz and I ended up at just logging in https://github.com/crossplane/crossplane-runtime/pull/754/commits/f7fb3a733a63a95ba0d4eb5f5555105d247dabd8